legend-engine-ide-lsp-default-extensions/src/main/java/org/finos/legend/engine/ide/lsp/extension/notebook/PureBookLSPGrammarExtension.java

```diff
@@ -0,0 +1,323 @@
+        ).whenCompleteAsync((l, e) -> this.generatePlan(sectionState), sectionState.getDocumentState().getGlobalState().getForkJoinPool());
+    }
+
+    @Override
+    public Iterable<? extends LegendDiagnostic> getDiagnostics(SectionState sectionState)
+    {
+        DocumentState documentState = sectionState.getDocumentState();
+        TextLocation sectionLocation = TextLocation.newTextSource(documentState.getDocumentId(), sectionState.getSection().getTextInterval());
+        if (!(documentState instanceof NotebookDocumentState))
+        {
+            return List.of(LegendDiagnostic.newDiagnostic(sectionLocation, "###purebook should not be use outside of Purebooks", LegendDiagnostic.Kind.Error, LegendDiagnostic.Source.Parser));
+        }
+
+        if (sectionState.getSection().getFullText().isEmpty())
+        {
+            return List.of();
+        }
+
+        try
+        {
+            try
+            {
+                this.compile(sectionState).get(30, TimeUnit.SECONDS);
+            }
+            catch (CompletionException | ExecutionException e)
+            {
+                throw e.getCause();
+            }
+        }
+        catch (EngineException e)
+        {
+            SourceInformation sourceInfo = e.getSourceInformation();
+            TextLocation location;
+            if (SourceInformationUtil.isValidSourceInfo(sourceInfo))
+            {
+                location = SourceInformationUtil.toLocation(sourceInfo);
+            }
+            else
+            {
+                location = sectionLocation;
+            }
+
+            LegendDiagnostic.Source source = e.getErrorType() == EngineErrorType.PARSER ? LegendDiagnostic.Source.Parser : LegendDiagnostic.Source.Compiler;
+
+            return List.of(LegendDiagnostic.newDiagnostic(location, e.getMessage(), LegendDiagnostic.Kind.Error, source));
+        }
+        catch (Throwable t)
+        {
+            return List.of(LegendDiagnostic.newDiagnostic(sectionLocation, t.getMessage(), LegendDiagnostic.Kind.Error, LegendDiagnostic.Source.Compiler));
+        }
+
+        return List.of();
+    }
+
+    @Override
+    public Stream<LegendReference> getLegendReferences(SectionState sectionState)
+    {
+        CompletableFuture<? extends LambdaFunction<?>> compiled = this.compile(sectionState);
+        try
+        {
+            LambdaFunction<?> lambdaFunction = compiled.get(30, TimeUnit.SECONDS);
+            PureModel pureModel = this.pureGrammarExtension.getCompileResult(sectionState).getPureModel();
+            return LegendReferenceResolver.toLegendReference(
+                    sectionState,
+                    FUNCTION_EXPRESSION_NAVIGATOR.findReferences(Optional.ofNullable(lambdaFunction)),
+                    pureModel.getContext()
+            );
+        }
+        catch (Exception ignore)
+        {
+            // ignore for now...
+        }
```

Comment:
Shouldn't this be logged?

Reply:
Yes... will fix it on my next set of changes.